### PR TITLE
Stop the molab dependency parser at shell command separators

### PR DESCRIPTION
**Pull request by danielhanchen**
## Problem

`scripts/molab_dependencies.py` extracts the package list for the generated `molab/*.py` PEP 723 headers from the install cells of `nb/*.ipynb`. `_RE_PIP_INSTALL` matches to the end of the line, so a line that chains two shell commands arrives at the token loop whole:

```
!pip install transformers==4.55.4 && pip install --no-deps trl==0.22.2
```

The loop skipped the `&&` token but kept parsing afterwards, so `pip` and `install`, which are the next command's name and subcommand, matched `_RE_PKG_NAME` and were emitted as package specs.

The result is visible in `molab/Qwen3_(4B)_Instruct-QAT.py`:

```
# dependencies = [
#     "huggingface_hub>=0.34.0",
#     "install",
#     "marimo",
#     "peft",
#     "pip",
```

`install` is an unrelated third party project on PyPI and it was requested with no version pin, so uv resolves and installs it whenever the notebook environment is created. Installing a package can execute code through build hooks or `.pth` files, and a molab runtime usually holds the user's Hugging Face token and their dataset, so an unintended name in that list is a supply chain path we should not be handing out.

## Fix

Split the argument tokens on `&&`, `||`, `;`, `|` and `&`, then take arguments only from segments that are themselves a pip install (`pip install`, `uv pip install`, `pip3 install`, `python -m pip install`).

The subtlety worth calling out: simply stopping at `&&` would have silently dropped `trl==0.22.2`, which is a pin the notebook genuinely needs. Segmenting keeps the second command's packages and drops only its command words. A chained non pip command such as `echo ...` contributes nothing.

A short `_NEVER_A_PACKAGE` list of bare shell words (`pip`, `install`, `python`, `echo`, `apt` and friends) is applied as a second layer, so a future install line shape that gets past the segmenter still cannot put a command word in a dependency block. Only exact unversioned tokens are listed, so a real pin such as `pip==25.0` would still be honoured.

## Blast radius

I ran `plan_dependencies()` over all 442 notebooks in `nb/` before and after the change and diffed the results. Exactly 2 notebooks change, `Qwen3_(4B)_Instruct-QAT` and its Kaggle twin, and each one only loses `"install"` and `"pip"`. The other 440 are identical.

Nothing user facing moves. The source notebooks in `nb/` are untouched, so Colab and Kaggle flows are unchanged, and the QAT molab notebook still installs `transformers==4.55.4`, `trl==0.22.2`, `torchao>=0.16.0` and the rest of its real stack.

## Idempotency

`molab/Qwen3_(4B)_Instruct-QAT.py` is committed here with those two lines removed. Regenerating it with the CI pins (`marimo==0.23.8`, `ruff==0.15.16`) produces a byte identical file, and `python scripts/molab_generate.py --check` exits 0 across the whole catalog, so re-running the generator is a no-op.

## Tests

Added to `tests/test_molab_dependency_parity.py`, so no new CI registration is needed:

- `test_parse_pip_line_drops_chained_command_words`: the exact chained line from the QAT notebook keeps both packages and neither command word.
- `test_parse_pip_line_ignores_chained_non_pip_command`: a chained `echo ...` contributes nothing.
- `test_no_shell_command_word_as_dependency`: catalog wide guard over every generated PEP 723 block, so this class of leak fails CI rather than shipping.

Suites run locally: `test_molab_dependency_parity.py` (400 passed), `test_molab_generation_parity.py`, `test_molab_templated_pins.py`, `test_molab_manifest.py`, `test_molab_no_colab_markers.py`, `test_molab_readme_links.py`, `test_every_test_runs_in_ci.py` (2346 passed), `test_molab_marimo_validity.py` (840 passed). Full `pytest tests` run is 16643 passed, 569 skipped, with 4 pre-existing failures in `test_notebook_pin_consistency.py` on the Muse Glimmer 30B GRPO notebooks that reproduce identically on `main` without this branch.

### CODE REVIEW

**Comment by danielhanchen** [comment]
@codex review

File: tests/test_molab_dependency_parity.py [hunk 1]
```diff
@@ -287,6 +287,144 @@ def _import_dependencies():
...
+
+    nb = _STEM_TO_NB.get(py_file.stem)
+    nb_name = nb.source.name if nb is not None else f"{py_file.stem}.ipynb"
+    source_nb_path = REPO_ROOT / "nb" / nb_name
+    if not source_nb_path.exists():
+        pytest.skip(f"Source notebook {nb_name} not found.")
+
+    packages = _extract_pep723_packages(py_file.read_text(encoding="utf-8"))
+    requested = _extract_pip_packages_from_nb(source_nb_path)
+    leaked = sorted(packages & _SHELL_COMMAND_WORDS - requested)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_molab_dependency_parity.py:409]: Not taking this one. The observation about `_NON_PACKAGE_TOKENS` is accurate, but the remedy costs more than the gap.

The plausible case was `sh`, and that already works: `sh` is not in `_NON_PACKAGE_TOKENS`, so a notebook pinning it lands in `requested` and the guard stays quiet. What remains is only the names that are also pip's own CLI words, so the guard would need a notebook to install a distribution literally called `pip` or `install` before it fired. No notebook does, and if one ever did I would rather CI ask a human than have the check look the other way on the exact two names this PR exists to keep out.

The fix implied here is to stop filtering those words in `_extract_pip_packages_from_nb`, and that helper is shared with `test_dependency_parity`, which asserts every source package appears in the generated file. 98 notebooks in nb/ have install lines whose tokens include `pip`, `install` or `uv` (`!pip install --upgrade -qqq uv`, chained commands, and so on), so those words would immediately become required dependencies for all of them and the parity test would fail across the catalog. Trading a live, catalog-wide test for an unreachable edge case is the wrong direction.

Leaving the guard as it is.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review
